scipy/special/_basic.py

```diff
@@ -2791,6 +2791,12 @@ def perm(N, k, exact=False):
 
     """
     if exact:
+        if not (isscalar(N) and isscalar(k)):
+            raise ValueError("`N` and `k` must scalar integers be with `exact=True`.")
+        if not (floor(N) == N and floor(k) == k):
```

Comment:
Sorry if this has already been discussed, but are we going to continue accepting integral float scalars?
Currently,

```python3
from scipy import special
special.perm(100., 50., exact=True)
```
just raises an error but no warning. ISTM that we should either not accept the floats or, since we are already taking `floor`, we might as well convert them to `int`s and use the result.

Reply:
It does not really raise an error (as in: intentionally), the implementation simply isn't robust for that case and falls on its face. Going forward, I'm in favour of deprecating `exact=True` also for floats that satisfy `floor(N) == N`, but in the meantime we should definitely avoid splats coming from our implementation.

IOW, things that fail already should just be turned into an outright error (rather than deprecating hypothetical usage that's _already_ broken). We should only deprecate cases that currently produce output (whether correct or incorrect).

Reply:
Ah, I suppose I assumed that before this PR, `perm` with `exact=True` and non-integral arguments behaved like `comb` and always produced results as if `exact=False`. I see the original and new behavior now.

Reply:
@h-vetinari I couldn't quite tell if you wanted me to change something in this PR?

Reply:
Thanks for double-checking. Indeed, my point is that code that is already failing in numpy 1.14 and before _should not be deprecated_, but directly turned into an error with an informative message.

AFAICT from some simple testing, anything that doesn't trigger the conditions already in place for an early return _will_ fail for `exact=True` if floats are involved. Those things should directly raise an error. The rest that doesn't fail can be deprecated.